localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -375,42 +380,25 @@ void NDTScanMatcher::callback_initial_pose(
 {
   if (!is_activated_) return;
 
-  // lock mutex for initial pose
-  std::lock_guard<std::mutex> initial_pose_array_lock(initial_pose_array_mtx_);
-  // if rosbag restart, clear buffer
-  if (!initial_pose_msg_ptr_array_.empty()) {
-    const builtin_interfaces::msg::Time & t_front =
-      initial_pose_msg_ptr_array_.front()->header.stamp;
-    const builtin_interfaces::msg::Time & t_msg = initial_pose_msg_ptr->header.stamp;
-    if (t_front.sec > t_msg.sec || (t_front.sec == t_msg.sec && t_front.nanosec > t_msg.nanosec)) {
-      initial_pose_msg_ptr_array_.clear();
-    }
+  if (initial_pose_buffer_->detect_time_jump_to_past(initial_pose_msg_ptr->header.stamp)) {
```

Comment:
Please explain why this check exists. We should make sure that future viewers of this code will know why this process is necessary. In my understand, if the system is normal, the clock should not go back to the past.
e.g. `// This is to ensure that if rosbag is restarted, the process will continue.`

Reply:
@YamatoAndo @KYabuuchi 
Is it necessary to take measures against rosbag replay?
I would like to make sure that there is motivation to deal with this, since it is not only callback_initial_pose, but also the node as a whole (or rather Autoware as a whole) that needs to be addressed to make it normal.
(I would start by somehow building up a test to simulate such a replay if necessary)

https://github.com/autowarefoundation/autoware.universe/blob/e3d4a1854e9786b72694a7a2b7716a040da98d24/localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp#L383-L385


I am beginning to feel that it would be OK to remove the detect_time_jump_to_past function and check the alignment of the time at push funcion.

Reply:
Removed detect_time_jump_to_past function and check in push function instead
https://github.com/autowarefoundation/autoware.universe/pull/5713/commits/663812bd93daa486471450f35836ad86e24d4965